airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java

```diff
@@ -361,11 +367,27 @@ public WebBackendConnectionRead webBackendUpdateConnection(final WebBackendConne
     connectionRead = connectionsHandler.updateConnection(connectionUpdate);
 
     if (needReset) {
+      final UUID connectionId = webBackendConnectionUpdate.getConnectionId();
+      final ConnectionIdRequestBody connectionIdRequestBody = new ConnectionIdRequestBody().connectionId(connectionId);
+      final ConnectionStateType stateType = getStateType(connectionIdRequestBody);
+      final List<io.airbyte.protocol.models.StreamDescriptor> streamsToReset;
+
+      if (stateType == ConnectionStateType.LEGACY || stateType == ConnectionStateType.NOT_SET) {
+        streamsToReset = configRepository.getAllStreamsForConnection(connectionId);
+      } else {
+        final ConfiguredAirbyteCatalog existingConfiguredCatalog =
+            configRepository.getConfiguredCatalogForConnection(connectionId);
+        final io.airbyte.protocol.models.AirbyteCatalog existingCatalog = CatalogHelpers.configuredCatalogToCatalog(existingConfiguredCatalog);
+        final AirbyteCatalog apiExistingCatalog = CatalogConverter.toApi(existingCatalog);
+        final AirbyteCatalog newAirbyteCatalog = webBackendConnectionUpdate.getSyncCatalog();
+        final CatalogDiff catalogDiff = connectionsHandler.getDiff(apiExistingCatalog, newAirbyteCatalog);
+        final List<StreamDescriptor> apiStreamsToReset = getStreamsToReset(catalogDiff);
+        streamsToReset = apiStreamsToReset.stream().map(streamDescriptor -> ProtocolConverters.streamDescriptorToProtocol(streamDescriptor)).toList();
```

Comment:
nit: single-argument lambdas like this can be slightly simplified like so:
```suggestion
        streamsToReset = apiStreamsToReset.stream().map(ProtocolConverters::streamDescriptorToProtocol).toList();
```